### ISSUE TITLE
With lsp-organize-imports as a before-save-hook, saving immediately after editing lags for a 0.5-1 seconds

### ISSUE THREAD

**Issue by jimeh** [emacs-lsp/lsp-mode#2386]
**Describe the bug**

When using `lsp-organize-imports` as a before-save hook, after PR #2025 triggering a save immediately after modifying buffer content causes a slight delay/lag with applying the formatting changes. While if you make a change, wait around 1 second and trigger a save, the formatting changes apply instantly.

I have narrowed it down to commit a2896d5b (#2025), the commit before it (c4daa529) does not exhibit this behavior, and instead applies the formatting changes instantly without any delay regardless of the amount of time you wait between modifying buffer content and triggering save.

Below is a gif of the bug in action with lsp-mode on commit a2896d5b. Notice the delay after `C-x C-s` (`^X` `^S` in the keystroke visualizations) and when the formatting change from `gopls` is applied if I do not wait between `C-o` and `C-x C-s` commands. While when I wait about a second, it applies the formatting change instantly.

![lsp-mode-save-lag-after-pr-2025](https://user-images.githubusercontent.com/39563/101428176-1fb60700-38f8-11eb-966e-841f62014641.gif)

And here is a gif with lsp-mode on the commit right before (c4daa529) where there is no delay with applying formatting changes:

![lsp-mode-save-lag-before-pr-2025](https://user-images.githubusercontent.com/39563/101428331-673c9300-38f8-11eb-8332-06ebf42687cd.gif)

**To Reproduce**

My personal Emacs config for lsp-mode+go is available [here](https://github.com/jimeh/.emacs.d/blob/9254df7905177efb72e9f834bb5ef4d62a0f0ef7/modules/languages/siren-golang.el#L70-L89).

However I have also re-produced the same behavior using the `lsp-start-plain.el` file with minor tweaks to enable formatting on save for `go-mode`, and confirmed it does not occur with lsp-mode 7.0.1 from melpa-stable. Below are the two variants of the start-plain file I used:

<details>
  <summary><code>lsp-start-plain.el</code></summary>

```elisp
;;; lsp-start-plain.el --- LSP mode quick starter      -*- lexical-binding: t; -*-

;; Copyright (C) 2018 Ivan Yonchovski

;; Author: Zhu Zihao <all_but_last@163.com>
;; Keywords: languages

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; This file is a helper to start a minimal lsp environment.
;; To use this, start your Emacs with "emacs -q" and load this file.

;; It will install `lsp-mode', `lsp-ui' with their dependencies to start a
;; minimal lsp environment.

;; And it forces Emacs to load `.el' files rather than `.elc' files
;; for more readable backtrace.

;;; Code:

(require 'package)

(setq debug-on-error t)

(let* ((package-archives '(("melpa" . "https://melpa.org/packages/")
                           ("gnu" . "https://elpa.gnu.org/packages/")))
       (no-byte-compile t)
       (package-user-dir (expand-file-name (make-temp-name "lsp-tmp-elpa")
                                           user-emacs-directory))
       (custom-file (expand-file-name "custom.el" package-user-dir))
       (pkg-list '(lsp-mode company flycheck go-mode)))

  (package-initialize)
  (package-refresh-contents)

  (mapcar (lambda (pkg)
            (unless (package-installed-p pkg)
              (package-install pkg))
            (require pkg))
          pkg-list)

  (add-hook 'prog-mode-hook 'lsp)

  (defun lsp-go-install-save-hooks ()
    (add-hook 'before-save-hook #'lsp-format-buffer t t)
    (add-hook 'before-save-hook #'lsp-organize-imports t t))
  (add-hook 'go-mode-hook #'lsp-go-install-save-hooks)

  (add-hook 'kill-emacs-hook `(lambda ()
                                (delete-directory ,package-user-dir t))))

(provide 'lsp-start-plain)
;;; lsp-start-plain.el ends here
```

</details>
<details>
  <summary><code>lsp-start-plain.stable.el</code></summary>

```elisp
;;; lsp-start-plain.stable.el --- LSP mode quick starter      -*- lexical-binding: t; -*-

;; Copyright (C) 2018 Ivan Yonchovski

;; Author: Zhu Zihao <all_but_last@163.com>
;; Keywords: languages

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; This file is a helper to start a minimal lsp environment.
;; To use this, start your Emacs with "emacs -q" and load this file.

;; It will install `lsp-mode', `lsp-ui' with their dependencies to start a
;; minimal lsp environment.

;; And it forces Emacs to load `.el' files rather than `.elc' files
;; for more readable backtrace.

;;; Code:

(require 'package)

(setq debug-on-error t)

(let* ((package-archives '(("melpa-stable" . "https://stable.melpa.org/packages/")
                           ("gnu" . "https://elpa.gnu.org/packages/")))
       (no-byte-compile t)
       (package-user-dir (expand-file-name (make-temp-name "lsp-tmp-elpa")
                                           user-emacs-directory))
       (custom-file (expand-file-name "custom.el" package-user-dir))
       (pkg-list '(lsp-mode company flycheck go-mode)))

  (package-initialize)
  (package-refresh-contents)

  (mapcar (lambda (pkg)
            (unless (package-installed-p pkg)
              (package-install pkg))
            (require pkg))
          pkg-list)

  (add-hook 'prog-mode-hook 'lsp)

  (defun lsp-go-install-save-hooks ()
    (add-hook 'before-save-hook #'lsp-format-buffer t t)
    (add-hook 'before-save-hook #'lsp-organize-imports t t))
  (add-hook 'go-mode-hook #'lsp-go-install-save-hooks)

  (add-hook 'kill-emacs-hook `(lambda ()
                                (delete-directory ,package-user-dir t))))

(provide 'lsp-start-plain)
;;; lsp-start-plain.stable.el ends here
```

</details>

The project/file I've been using to test this, is [`undent.go`](https://github.com/jimeh/undent/blob/d79e413e8e850311b314af5895e9a0e30330edd9/undent.go) from my own super-simple [undent](https://github.com/jimeh/undent) project.

**Expected behavior**

There to be no lag like before #2025 was merged.

**Which Language Server did you use**

I'm using Golang's gopls language server version 0.5.5:

```
$ gopls version
golang.org/x/tools/gopls master
    golang.org/x/tools/gopls@v0.5.5 h1:YBLs2WCwxx6mBF98Y6/mDIpKhP4Dt4Dz2q6LnMRBpDw=
```

**OS**

- macOS Catalina 10.15.7 (19H15)
- gccemacs / native-comp, built from commit [7a8370e](https://github.com/emacs-mirror/emacs/commit/7a8370ed0f1b1d62657e385789ee2f81c5607ec5) with my own [build script](https://github.com/jimeh/build-emacs-for-macos).

**Error callstack**

I don't have a error callstack, as there's no error, but I did run a CPU profiler, while running `C-o C-o C-x C-s` just like in the gifs above.

<details>
  <summary><code>lsp-cpu-report-after-pr-2025-a2896d5b.txt</code></summary>

```
- ...                                                              35  58%
   Automatic GC                                                    26  43%
 - lsp-modeline--check-code-actions                                 7  11%
  - if                                                              7  11%
   - progn                                                          7  11%
    - save-excursion                                                6  10%
     - let                                                          6  10%
      - if                                                          6  10%
       - progn                                                      6  10%
        - if                                                        5   8%
         - condition-case                                           5   8%
          - flycheck-buffer                                         5   8%
           - if                                                     5   8%
            - if                                                    5   8%
             - condition-case                                       5   8%
              - let*                                                5   8%
               - if                                                 5   8%
                - progn                                             4   6%
                 - lsp-request-async                                2   3%
                  - let*                                            2   3%
                   - progn                                          2   3%
                    - lsp--send-request-async                       2   3%
                     - let*                                         2   3%
                      - if                                          2   3%
                       - let*                                       2   3%
                        - seq-do                                    2   3%
                         - mapc                                     2   3%
                          - #<lambda -0x1dc2b435a7f9dfab>                  2   3%
                           - let                                    2   3%
                            - lsp--send-no-wait                     2   3%
                               condition-case                       2   3%
                 - let*                                             2   3%
                  - lsp-glob-to-regexp                              2   3%
                   - concat                                         2   3%
                    - lsp--glob-to-regexp                           2   3%
                     - if                                           2   3%
                      - let                                         2   3%
                       - lsp-split-glob-pattern                     2   3%
                        - if                                        2   3%
                         - progn                                    2   3%
                          - let                                     2   3%
                           - let                                    2   3%
                            - funcall                               1   1%
                             - #<lambda 0x53cd2b640258535>                  1   1%
                              - car                                 1   1%
                               - -reduce-r-from                     1   1%
                                - let                               1   1%
                                 - let                              1   1%
                                  - while                           1   1%
                                   - let                            1   1%
                                    - setq                          1   1%
                                     - funcall                      1   1%
                                      - #<lambda 0x2dec558e8f142de>                  1   1%
                                       - list                       1   1%
                                        - apply                     1   1%
                                           lsp--workspace-client                  1   1%
                            - while                                 1   1%
                               let                                  1   1%
                - flycheck-start-current-syntax-check                  1   1%
                 - let*                                             1   1%
                  - flycheck-syntax-check-new                       1   1%
                   - flycheck-compute-working-directory                  1   1%
                    - let*                                          1   1%
                       if                                           1   1%
        - setq                                                      1   1%
         - or                                                       1   1%
          - go--in-interface-p                                      1   1%
           - go--in-paren-with-prefix-p                             1   1%
            - save-excursion                                        1   1%
             - and                                                  1   1%
              - go-goto-opening-parenthesis                         1   1%
               - let                                                1   1%
                - nth                                               1   1%
                   syntax-ppss                                      1   1%
    - lsp-request-async                                             1   1%
     - let*                                                         1   1%
      - progn                                                       1   1%
       - lsp--send-request-async                                    1   1%
        - let*                                                      1   1%
         - if                                                       1   1%
          - let*                                                    1   1%
           - seq-do                                                 1   1%
            - mapc                                                  1   1%
             - #<lambda -0x3f5a1489b1fafab>                         1   1%
              - let                                                 1   1%
               - lsp--make-message                                  1   1%
                  let                                               1   1%
 - lsp-organize-imports                                             2   3%
  - condition-case                                                  2   3%
   - lsp-execute-code-action-by-kind                                2   3%
    - let*                                                          2   3%
     - and                                                          2   3%
      - lsp--select-action                                          2   3%
       - -filter                                                    2   3%
        - lsp-get-or-calculate-code-actions                         2   3%
         - lsp-request                                              2   3%
          - let*                                                    2   3%
           - progn                                                  2   3%
            - if                                                    2   3%
             - let*                                                 2   3%
              - unwind-protect                                      2   3%
               - progn                                              2   3%
                - lsp-request-async                                 1   1%
                 - let*                                             1   1%
                  - progn                                           1   1%
                   - lsp--send-request-async                        1   1%
                    - let*                                          1   1%
                     - if                                           1   1%
                      - let*                                        1   1%
                       - seq-do                                     1   1%
                        - mapc                                      1   1%
                         - #<lambda 0x1f09860a301cd146>                  1   1%
                          - let                                     1   1%
                           - lsp--send-no-wait                      1   1%
                              condition-case                        1   1%
                  while                                             1   1%
- command-execute                                                  21  35%
 - funcall-interactively                                           14  23%
  - save-buffer                                                    11  18%
   - basic-save-buffer                                             11  18%
    - run-hooks                                                    10  16%
     - lsp-format-buffer                                           10  16%
      - cond                                                       10  16%
       - let                                                       10  16%
        - lsp-request                                               9  15%
         - let*                                                     9  15%
          - progn                                                   9  15%
           - if                                                     9  15%
            - let*                                                  9  15%
             - unwind-protect                                       9  15%
              - progn                                               9  15%
                 while                                              9  15%
        - if                                                        1   1%
         - lsp--apply-text-edits                                    1   1%
          - if                                                      1   1%
           - let                                                    1   1%
            - unwind-protect                                        1   1%
             - progn                                                1   1%
              - prog1                                               1   1%
               - progn                                              1   1%
                - let*                                              1   1%
                   make-progress-reporter                           1   1%
    - basic-save-buffer-1                                           1   1%
       basic-save-buffer-2                                          1   1%
  - execute-extended-command                                        3   5%
   - command-execute                                                3   5%
      funcall-interactively                                         3   5%
 - byte-code                                                        7  11%
  - read-extended-command                                           7  11%
     completing-read-default                                        5   8%
- redisplay_internal (C function)                                   4   6%
 - eval                                                             2   3%
    mode-line-eol-desc                                              1   1%
  - mapconcat                                                       1   1%
   - lsp--workspace-print                                           1   1%
    - let*                                                          1   1%
       propertize                                                   1   1%
 - jit-lock-function                                                2   3%
  - jit-lock-fontify-now                                            2   3%
   - jit-lock--run-functions                                        2   3%
    - #<compiled 0x5872cb788801ef5>                                 2   3%
     - font-lock-fontify-region                                     2   3%
      - font-lock-default-fontify-region                            2   3%
       - font-lock-fontify-keywords-region                          2   3%
        - go--match-decl                                            2   3%
         - let                                                      2   3%
          - while                                                   2   3%
           - save-excursion                                         2   3%
            - cond                                                  2   3%
             - and                                                  1   1%
              - setq                                                1   1%
               - go--containing-decl                                1   1%
                - let                                               1   1%
                 - unwind-protect                                   1   1%
                  - progn                                           1   1%
                   - or                                             1   1%
                    - save-excursion                                1   1%
                     - let                                          1   1%
                      - car                                         1   1%
                         syntax-ppss                                1   1%
               let                                                  1   1%
```

</details>

<details>
  <summary><code>lsp-cpu-report-before-pr-2025-c4daa529.txt</code></summary>

```
- command-execute                                                  37  51%
 - byte-code                                                       23  31%
  - read-extended-command                                          23  31%
   - completing-read-default                                       15  20%
    - command-execute                                               3   4%
     - funcall-interactively                                        3   4%
      - minibuffer-complete                                         3   4%
       - completion-in-region                                       3   4%
        - completion--in-region                                     3   4%
         - #<compiled -0x1a89e7f44c4431eb>                          3   4%
          - apply                                                   3   4%
           - #<compiled 0x2ba11fb3c3318fb>                          3   4%
            - completion--in-region-1                               3   4%
             - completion--do-completion                            3   4%
              - completion-try-completion                           3   4%
               - completion--nth-completion                         3   4%
                - completion--some                                  3   4%
                 - #<compiled -0x1ed2974c01ff2297>                  2   2%
                  - completion-basic-try-completion                  2   2%
                   - #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_42>                  2   2%
                      complete-with-action                          2   2%
 - funcall-interactively                                           14  19%
  - save-buffer                                                     6   8%
   - basic-save-buffer                                              6   8%
    - run-hooks                                                     3   4%
     - lsp-format-buffer                                            3   4%
      - cond                                                        3   4%
       - let                                                        2   2%
        - if                                                        2   2%
         - setq                                                     1   1%
          - funcall                                                 1   1%
           - #<lambda -0x129e1a7f9ce25d63>                          1   1%
            - and                                                   1   1%
             - seq-find                                             1   1%
              - seq-do                                              1   1%
               - mapc                                               1   1%
                - #<compiled 0xcda58225f212d6f>                     1   1%
                 - #<lambda 0x146232dee9731436>                     1   1%
                  - let*                                            1   1%
                   - let*                                           1   1%
                    - or                                            1   1%
                     - string-match                                 1   1%
                        f-relative                                  1   1%
         - lsp--apply-text-edits                                    1   1%
          - if                                                      1   1%
           - let                                                    1   1%
            - unwind-protect                                        1   1%
             - progn                                                1   1%
              - prog1                                               1   1%
               - progn                                              1   1%
                - let*                                              1   1%
                   lsp--info                                        1   1%
       - lsp-feature?                                               1   1%
        - let                                                       1   1%
         - let                                                      1   1%
          - if                                                      1   1%
           - let                                                    1   1%
            - let*                                                  1   1%
             - let                                                  1   1%
              - let                                                 1   1%
               - while                                              1   1%
                - let                                               1   1%
                   if                                               1   1%
    - basic-save-buffer-1                                           2   2%
       basic-save-buffer-2                                          2   2%
    - vc-after-save                                                 1   1%
     - vc-state-refresh                                             1   1%
      - vc-call-backend                                             1   1%
       - vc-git-state                                               1   1%
        - vc-git--run-command-string                                1   1%
         - vc-git--out-ok                                           1   1%
          - vc-git--call                                            1   1%
             process-file                                           1   1%
  - execute-extended-command                                        5   6%
     sit-for                                                        4   5%
   - command-execute                                                1   1%
      funcall-interactively                                         1   1%
  - open-line                                                       3   4%
   - newline                                                        3   4%
    - flycheck-handle-change                                        3   4%
     - let                                                          3   4%
      - unwind-protect                                              3   4%
       - progn                                                      3   4%
        - if                                                        3   4%
         - progn                                                    3   4%
          - if                                                      3   4%
           - condition-case                                         2   2%
            - flycheck-buffer                                       2   2%
             - if                                                   2   2%
              - if                                                  2   2%
               - condition-case                                     2   2%
                - let*                                              2   2%
                 - flycheck-get-checker-for-buffer                  1   1%
                  - if                                              1   1%
                   - if                                             1   1%
                    - flycheck-may-use-checker                      1   1%
                     - let                                          1   1%
                      - and                                         1   1%
                       - or                                         1   1%
                        - funcall                                   1   1%
                         - #<lambda -0x21a1a53259d2a43>                  1   1%
                          - let                                     1   1%
                           - flycheck-compute-working-directory                  1   1%
                            - let*                                  1   1%
                               if                                   1   1%
                 - if                                               1   1%
                  - flycheck-start-current-syntax-check                  1   1%
                   - let*                                           1   1%
                    - flycheck-syntax-check-new                     1   1%
                     - flycheck-compute-working-directory                  1   1%
                      - let*                                        1   1%
                         if                                         1   1%
           - flycheck-buffer-automatically                          1   1%
            - if                                                    1   1%
             - and                                                  1   1%
              - flycheck-may-check-automatically                    1   1%
                 and                                                1   1%
- ...                                                              32  44%
   Automatic GC                                                    28  38%
 - lsp--document-links                                              3   4%
  - if                                                              3   4%
   - progn                                                          3   4%
    - lsp-request-async                                             3   4%
     - let*                                                         3   4%
      - progn                                                       2   2%
       - lsp--send-request-async                                    2   2%
        - let*                                                      2   2%
         - if                                                       2   2%
          - let*                                                    2   2%
           - seq-do                                                 2   2%
            - mapc                                                  2   2%
             - #<lambda -0x193a054025629d8c>                        2   2%
              - let                                                 2   2%
               - or                                                 1   1%
                - if                                                1   1%
                 - progn                                            1   1%
                  - lsp--capability                                 1   1%
                   - lsp-get                                        1   1%
                    - or                                            1   1%
                     - lsp--server-capabilities                     1   1%
                      - apply                                       1   1%
                       - lsp-merge                                  1   1%
                        - let                                       1   1%
                         - mapc                                     1   1%
                          - #<lambda 0x23fa97160697a5>                  1   1%
                           - ht-update!                             1   1%
                            - maphash                               1   1%
                               #<lambda 0x33b0ca301>                  1   1%
               - lsp--send-no-wait                                  1   1%
                  condition-case                                    1   1%
      - unwind-protect                                              1   1%
       - progn                                                      1   1%
        - lsp-request-async                                         1   1%
         - let*                                                     1   1%
          - progn                                                   1   1%
           - lsp--send-request-async                                1   1%
            - let*                                                  1   1%
             - if                                                   1   1%
              - let*                                                1   1%
               - seq-do                                             1   1%
                - mapc                                              1   1%
                 - #<lambda 0x1f1804698456c876>                     1   1%
                  - let                                             1   1%
                   - lsp--send-no-wait                              1   1%
                      condition-case                                1   1%
 - #<lambda -0x173f0895cc56cb2c>                                    1   1%
  - funcall                                                         1   1%
   - lsp--apply-text-edit-replace-buffer-contents                   1   1%
    - let*                                                          1   1%
     - let*                                                         1   1%
      - let                                                         1   1%
       - save-current-buffer                                        1   1%
        - unwind-protect                                            1   1%
         - progn                                                    1   1%
          - let                                                     1   1%
           - save-current-buffer                                    1   1%
            - save-excursion                                        1   1%
             - save-restriction                                     1   1%
              - let                                                 1   1%
               - run-hook-with-args                                 1   1%
                  lsp-before-change                                 1   1%
- redisplay_internal (C function)                                   2   2%
   eval                                                             1   1%
 - jit-lock-function                                                1   1%
  - jit-lock-fontify-now                                            1   1%
   - jit-lock--run-functions                                        1   1%
    - #<compiled 0x5872cb78bd0def5>                                 1   1%
     - font-lock-fontify-region                                     1   1%
        font-lock-default-fontify-region                            1   1%
- flycheck-perform-deferred-syntax-check                            1   1%
 - if                                                               1   1%
  - progn                                                           1   1%
   - flycheck-buffer-automatically                                  1   1%
    - if                                                            1   1%
     - and                                                          1   1%
      - flycheck-may-check-automatically                            1   1%
         and                                                        1   1%
```

</details>

**Comment by yyoncho** [comment]
Thank you for the analysis. I really don't see how  a2896d5 could cause that issue unless it is related to the GC. If it generates a lot of garbage, then that would cause the freeze. I know that gopls sends some huge globs so that might be causing the memory explosion. But that freeze will happen non-deterministically, not only on save which is odd. Do you reproduce it with setup 100% of the time? Also, can you try setting gc-cons-threshold to 100mb? (we should update lsp-start-plain.el to have it as well)

In your config, I see that you set it once to max-positive-number: https://github.com/jimeh/.emacs.d/search?q=gc-cons-threshold - that should not happen as well(not sure if it is related).

**Comment by yyoncho** [comment]
cc @leungbk

**Comment by jimeh** [comment]
@yyoncho Yeah, I had a quick look at the commit and couldn’t quite understand what might be causing the delay. But I am so far able to reproduce it 100% of the time during the testing I did last night to put together this issue. I had noticed it on and off before then, but for now I’m assuming that’s due to the timing of when I saved.

With my personal config I took inspiration from doom-emacs earlier this year in an attempt to improve startup speed. So I set `gc-cons-threshold` to max in `early-init.el`, but ultimately use gchm to optimize GC, loaded as early as possible after the package system is ready [here](https://github.com/jimeh/.emacs.d/blob/9254df7905177efb72e9f834bb5ef4d62a0f0ef7/core/siren-core-performance.el), and with `gcmh-high-cons-threshold` set to 100MiB.

I also set `read-process-output-max` to 1MiB [here](https://github.com/jimeh/.emacs.d/blob/9254df7905177efb72e9f834bb5ef4d62a0f0ef7/modules/lsp/siren-lsp.el#L33).

I’ll try tweaking those numbers higher shortly when I’m back on my computer.

**Comment by jimeh** [comment]
After some more experimentation with the two start-plain files, I've made a few more discoveries:

- Setting both `gc-cons-threshold` and `read-process-output-max` to 1GiB (`(* 1024 1024 1024)`) does not help with the issue.
- I disabled the `lsp-organize-imports` before-save hook and setup a keybinding (`C-c C-f`) to run it manually, and discovered:
  - Delay happens only the first time `lsp-organize-imports` is triggered after the file is saved, and if the buffer content was modified immediately before.
  - After the initial trigger of `lsp-organize-imports`, all subsequent calls do not yield a delay in applying formatting changes regardless of time between last buffer content change and triggering of `lsp-organize-imports`.
  - Once the file is saved however, the delay happens again on first trigger, but only if the buffer content was modified immediately before. If you modify the buffer content, wait 1s and then trigger `lsp-organize-imports` changes are applied instantly.
  - After saving, you can wait and/or move cursor around for however long you want, first trigger of `lsp-organize-imports` will still cause a delay as long as it's immediately after modifying buffer content.

So seems these are the conditions that need to be met for the delay to happen:

- `lsp-organize-imports` must not have been executed since the last time the file was saved to disk.
- Buffer content must have been modified immediately before (less than 0.5-1s before) triggering `lsp-organize-imports`.

When manually triggering `lsp-organize-imports` via a keybinding, I found that I had to make changes that would cause the `imports` section of a go file to be updated. Either inserting extra blank lines within the `import (...)` that shouldn't be there, or adding/removing a `log.Fatal("")` or something similar to add/remove a new import line.

Below are the updated start-plain files I used:

<details>
<summary><code>lsp-start-plain.el</code></summary>

```elisp
;;; lsp-start-plain.el --- LSP mode quick starter      -*- lexical-binding: t; -*-

;; Copyright (C) 2018 Ivan Yonchovski

;; Author: Zhu Zihao <all_but_last@163.com>
;; Keywords: languages

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; This file is a helper to start a minimal lsp environment.
;; To use this, start your Emacs with "emacs -q" and load this file.

;; It will install `lsp-mode', `lsp-ui' with their dependencies to start a
;; minimal lsp environment.

;; And it forces Emacs to load `.el' files rather than `.elc' files
;; for more readable backtrace.

;;; Code:

(require 'package)

(setq debug-on-error t)
(setq gc-cons-threshold (* 1024 1024 1024))
(setq read-process-output-max (* 1024 1024 1024))

(let* ((package-archives '(("melpa" . "https://melpa.org/packages/")
                           ("gnu" . "https://elpa.gnu.org/packages/")))
       (no-byte-compile t)
       (package-user-dir (expand-file-name (make-temp-name "lsp-tmp-elpa")
                                           user-emacs-directory))
       (custom-file (expand-file-name "custom.el" package-user-dir))
       (pkg-list '(lsp-mode company flycheck go-mode)))

  (package-initialize)
  (package-refresh-contents)

  (mapcar (lambda (pkg)
            (unless (package-installed-p pkg)
              (package-install pkg))
            (require pkg))
          pkg-list)

  (add-hook 'prog-mode-hook 'lsp)

  (defun lsp-go-install-save-hooks ()
    (add-hook 'before-save-hook #'lsp-format-buffer t t)
    ;; (add-hook 'before-save-hook #'lsp-organize-imports t t)
    (define-key go-mode-map (kbd "C-c C-f") 'lsp-organize-imports))
  (add-hook 'go-mode-hook #'lsp-go-install-save-hooks)

  (add-hook 'kill-emacs-hook `(lambda ()
                                (delete-directory ,package-user-dir t))))

(provide 'lsp-start-plain)
;;; lsp-start-plain.el ends here
```

</details>
<details>
<summary><code>lsp-start-plain.stable.el</code></summary>

```elisp
;;; lsp-start-plain.stable.el --- LSP mode quick starter      -*- lexical-binding: t; -*-

;; Copyright (C) 2018 Ivan Yonchovski

;; Author: Zhu Zihao <all_but_last@163.com>
;; Keywords: languages

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; This file is a helper to start a minimal lsp environment.
;; To use this, start your Emacs with "emacs -q" and load this file.

;; It will install `lsp-mode', `lsp-ui' with their dependencies to start a
;; minimal lsp environment.

;; And it forces Emacs to load `.el' files rather than `.elc' files
;; for more readable backtrace.

;;; Code:

(require 'package)

(setq debug-on-error t)
(setq gc-cons-threshold (* 1024 1024 1024))
(setq read-process-output-max (* 1024 1024 1024))

(let* ((package-archives '(("melpa-stable" . "https://stable.melpa.org/packages/")
                           ("gnu" . "https://elpa.gnu.org/packages/")))
       (no-byte-compile t)
       (package-user-dir (expand-file-name (make-temp-name "lsp-tmp-elpa")
                                           user-emacs-directory))
       (custom-file (expand-file-name "custom.el" package-user-dir))
       (pkg-list '(lsp-mode company flycheck go-mode)))

  (package-initialize)
  (package-refresh-contents)

  (mapcar (lambda (pkg)
            (unless (package-installed-p pkg)
              (package-install pkg))
            (require pkg))
          pkg-list)

  (add-hook 'prog-mode-hook 'lsp)

  (defun lsp-go-install-save-hooks ()
    (add-hook 'before-save-hook #'lsp-format-buffer t t)
    ;; (add-hook 'before-save-hook #'lsp-organize-imports t t)
    (define-key go-mode-map (kbd "C-c C-f") 'lsp-organize-imports))
  (add-hook 'go-mode-hook #'lsp-go-install-save-hooks)

  (add-hook 'kill-emacs-hook `(lambda ()
                                (delete-directory ,package-user-dir t))))

(provide 'lsp-start-plain)
;;; lsp-start-plain.stable.el ends here
```

</details>

And I've also performed a few more CPU profiler runs. Though keep in mind that my invocation of `M-x report-stop` is probably mixed into these profiler reports.

<details>
<summary><code>lsp-immediate-organize-imports-first-call-after-pr-2025.txt</code></summary>

Delay occurred. After #2025, first invocation after file save, immediately after modifying buffer content.

```
- command-execute                                                  27  50%
 - byte-code                                                       16  30%
  - read-extended-command                                          16  30%
     completing-read-default                                       15  28%
 - funcall-interactively                                           11  20%
  - execute-extended-command                                       10  18%
     sit-for                                                        5   9%
   - command-execute                                                5   9%
      funcall-interactively                                         5   9%
  - open-line                                                       1   1%
   - newline                                                        1   1%
    - #<compiled 0x821938da5270781>                                 1   1%
     - move-to-left-margin                                          1   1%
      - current-left-margin                                         1   1%
       - back-to-indentation                                        1   1%
          internal--syntax-propertize                               1   1%
- ...                                                              25  47%
 - lsp-organize-imports                                            21  39%
  - condition-case                                                 21  39%
   - lsp-execute-code-action-by-kind                               21  39%
    - let*                                                         21  39%
     - and                                                         21  39%
      - lsp--select-action                                         21  39%
       - -filter                                                   21  39%
        - lsp-get-or-calculate-code-actions                        21  39%
         - lsp-request                                             21  39%
          - let*                                                   21  39%
           - progn                                                 21  39%
            - if                                                   21  39%
             - let*                                                21  39%
              - unwind-protect                                     21  39%
               - progn                                             21  39%
                - while                                            21  39%
                 - if                                               2   3%
                    <                                               1   1%
 - timer-event-handler                                              2   3%
  - apply                                                           2   3%
   - lsp--file-process-event                                        1   1%
    - let*                                                          1   1%
     - seq-do                                                       1   1%
      - mapc                                                        1   1%
       - #<lambda -0xfda73f16c6ea9b6>                               1   1%
        - if                                                        1   1%
         - progn                                                    1   1%
          - let                                                     1   1%
           - lsp-notify                                             1   1%
            - lsp--send-notification                                1   1%
             - mapcar                                               1   1%
              - #<lambda -0x2262e76d74f13cb>                        1   1%
               - let                                                1   1%
                - lsp--send-no-wait                                 1   1%
                   condition-case                                   1   1%
   - flycheck--handle-idle-trigger                                  1   1%
    - let                                                           1   1%
     - if                                                           1   1%
      - progn                                                       1   1%
       - save-current-buffer                                        1   1%
        - if                                                        1   1%
         - progn                                                    1   1%
          - flycheck-buffer-automatically                           1   1%
           - if                                                     1   1%
            - and                                                   1   1%
             - if                                                   1   1%
              - apply                                               1   1%
               - flycheck-may-check-automatically                   1   1%
                  and                                               1   1%
 - cond                                                             2   3%
  - let                                                             2   3%
   - let*                                                           2   3%
    - if                                                            2   3%
     - funcall                                                      1   1%
      - lsp--window-log-message                                     1   1%
       - let*                                                       1   1%
        - ignore                                                    1   1%
         - let                                                      1   1%
          - if                                                      1   1%
           - progn                                                  1   1%
            - lsp-log                                               1   1%
             - if                                                   1   1%
              - progn                                               1   1%
               - let                                                1   1%
                - save-current-buffer                               1   1%
                 - save-excursion                                   1   1%
                    let*                                            1   1%
     - save-current-buffer                                          1   1%
      - let*                                                        1   1%
       - if                                                         1   1%
        - lsp--apply-text-edits                                     1   1%
         - if                                                       1   1%
          - let                                                     1   1%
           - unwind-protect                                         1   1%
            - progn                                                 1   1%
             - prog1                                                1   1%
              - progn                                               1   1%
               - let*                                               1   1%
                  lsp--info                                         1   1%
   Automatic GC                                                     0   0%
  redisplay_internal (C function)                                   1   1%
```

</details>
<details>
<summary><code>lsp-immediate-organize-imports-second-call-fast-after-pr-2025.txt</code></summary>

Delay did not occurr. After #2025, second invocation after file save, immediately after modifying buffer content.

```
- command-execute                                                  21  60%
 - byte-code                                                       14  40%
  - read-extended-command                                          14  40%
     completing-read-default                                       14  40%
 - funcall-interactively                                            7  20%
  - execute-extended-command                                        6  17%
     sit-for                                                        3   8%
   - command-execute                                                2   5%
      funcall-interactively                                         2   5%
  - open-line                                                       1   2%
   - newline                                                        1   2%
    - flycheck-handle-change                                        1   2%
     - let                                                          1   2%
      - unwind-protect                                              1   2%
       - progn                                                      1   2%
        - if                                                        1   2%
         - progn                                                    1   2%
          - if                                                      1   2%
           - flycheck-buffer-automatically                          1   2%
            - if                                                    1   2%
             - and                                                  1   2%
              - if                                                  1   2%
               - flycheck-may-check-automatically                   1   2%
                  and                                               1   2%
- redisplay_internal (C function)                                  13  37%
 - jit-lock-function                                               12  34%
  - jit-lock-fontify-now                                           12  34%
   - jit-lock--run-functions                                       12  34%
    - #<compiled 0x5872e98fc15b1f5>                                12  34%
     - font-lock-fontify-region                                    12  34%
      - font-lock-default-fontify-region                           12  34%
       - font-lock-fontify-keywords-region                         11  31%
        - go--match-decl                                            9  25%
         - let                                                      9  25%
          - while                                                   9  25%
           - save-excursion                                         8  22%
            - cond                                                  8  22%
             - and                                                  5  14%
              - setq                                                5  14%
               - go--containing-decl                                5  14%
                - let                                               5  14%
                 - unwind-protect                                   5  14%
                  - progn                                           5  14%
                   - or                                             4  11%
                    - save-excursion                                4  11%
                     - and                                          3   8%
                      - go-goto-opening-parenthesis                  3   8%
                       - let                                        3   8%
                        - nth                                       3   8%
                           syntax-ppss                              3   8%
                     - let                                          1   2%
                      - and                                         1   2%
                       - =                                          1   2%
                        - car                                       1   2%
                           syntax-ppss                              1   2%
                   - lsp--check-document-changes-version                  1   2%
                    - if                                            1   2%
                     - seq-every-p                                  1   2%
                      - seq-do                                      1   2%
                       - mapc                                       1   2%
                        - #<compiled 0x14a738c85139df2f>                  1   2%
                         - #<lambda -0x517fa0ac169eb74>                  1   2%
                          - let*                                    1   2%
                           - or                                     1   2%
                            - let*                                  1   2%
                             - save-current-buffer                  1   2%
                              - set-buffer                          1   2%
                               - find-file-noselect                  1   2%
                                - file-truename                     1   2%
                                   file-truename                    1   2%
             - let                                                  3   8%
              - unwind-protect                                      3   8%
               - progn                                              3   8%
                - if                                                3   8%
                 - progn                                            3   8%
                  - let                                             3   8%
                   - setq                                           2   5%
                    - and                                           2   5%
                     - =                                            1   2%
                      - car                                         1   2%
                         syntax-ppss                                1   2%
                     - not                                          1   2%
                      - nth                                         1   2%
                         syntax-ppss                                1   2%
                   - if                                             1   2%
                    - progn                                         1   2%
                     - mapc                                         1   2%
                      - #<lambda -0xf4460daf43e24a5>                  1   2%
                       - let*                                       1   2%
                        - cond                                      1   2%
                         - let                                      1   2%
                          - lsp-notify                              1   2%
                           - lsp--send-notification                  1   2%
                            - mapcar                                1   2%
                             - #<lambda 0x7420e939ad483c>                  1   2%
                              - let                                 1   2%
                               - lsp--send-no-wait                  1   2%
                                  condition-case                    1   2%
             and                                                    1   2%
        - go--match-ident-colon                                     1   2%
         - let                                                      1   2%
          - while                                                   1   2%
           - save-excursion                                         1   2%
            - setq                                                  1   2%
             - or                                                   1   2%
              - and                                                 1   2%
               - or                                                 1   2%
                  eq                                                1   2%
        - go--fontify-param-pre                                     1   2%
         - setq                                                     1   2%
          - eq                                                      1   2%
           - go--parameter-list-type                                1   2%
            - save-excursion                                        1   2%
             - cond                                                 1   2%
                or                                                  1   2%
   eval                                                             1   2%
- ...                                                               1   2%
 - timer-event-handler                                              1   2%
  - apply                                                           1   2%
   - flycheck--handle-idle-trigger                                  1   2%
    - let                                                           1   2%
     - if                                                           1   2%
      - progn                                                       1   2%
       - save-current-buffer                                        1   2%
        - if                                                        1   2%
         - progn                                                    1   2%
          - flycheck-buffer-automatically                           1   2%
           - if                                                     1   2%
            - and                                                   1   2%
             - if                                                   1   2%
              - apply                                               1   2%
               - flycheck-may-check-automatically                   1   2%
                  and                                               1   2%
   Automatic GC                                                     0   0%
```

</details>
<details>
<summary><code>lsp-delayed-organize-imports-first-call-after-pr-2025.txt</code></summary>

Delay did not occurr. After #2025, first invocation after file save, about 1s after modifying buffer content.

```
- command-execute                                                  28  44%
 - byte-code                                                       17  26%
  - read-extended-command                                          17  26%
     completing-read-default                                       14  22%
 - funcall-interactively                                           11  17%
  - execute-extended-command                                        9  14%
   - sit-for                                                        6   9%
    - redisplay_internal (C function)                               1   1%
     - eval                                                         1   1%
      - mapconcat                                                   1   1%
       - lsp--workspace-print                                       1   1%
        - let*                                                      1   1%
           if                                                       1   1%
   - command-execute                                                3   4%
      funcall-interactively                                         3   4%
    open-line                                                       2   3%
- redisplay_internal (C function)                                  22  34%
 - jit-lock-function                                               22  34%
  - jit-lock-fontify-now                                           22  34%
   - jit-lock--run-functions                                       22  34%
    - #<compiled 0x5872e98fc15b1f5>                                22  34%
     - font-lock-fontify-region                                    22  34%
      - font-lock-default-fontify-region                           22  34%
       - font-lock-fontify-keywords-region                         18  28%
        - go--match-decl                                           13  20%
         - let                                                     13  20%
          - while                                                  12  19%
           - save-excursion                                        10  15%
            - cond                                                 10  15%
             - and                                                 10  15%
              - setq                                               10  15%
               - go--containing-decl                               10  15%
                - let                                              10  15%
                 - unwind-protect                                   7  11%
                  - progn                                           7  11%
                   - or                                             7  11%
                    - save-excursion                                6   9%
                     - let                                          4   6%
                      - car                                         2   3%
                         syntax-ppss                                2   3%
                      - and                                         2   3%
                       - =                                          2   3%
                        - car                                       2   3%
                           syntax-ppss                              2   3%
                     - and                                          2   3%
                      - go-goto-opening-parenthesis                  2   3%
                       - let                                        2   3%
                        - nth                                       2   3%
                           syntax-ppss                              2   3%
             and                                                    1   1%
          - cond                                                    1   1%
           - funcall                                                1   1%
            - #<compiled -0x139e466725aee51e>                       1   1%
             - apply                                                1   1%
              - lsp--file-process-event                             1   1%
               - let*                                               1   1%
                - seq-do                                            1   1%
                 - mapc                                             1   1%
                  - #<lambda -0xfda73f16c6ea9b6>                    1   1%
                   - if                                             1   1%
                    - progn                                         1   1%
                     - let                                          1   1%
                      - lsp-notify                                  1   1%
                       - lsp--send-notification                     1   1%
                          mapcar                                    1   1%
        - go--match-param-start                                     2   3%
         - let                                                      2   3%
          - while                                                   2   3%
           - if                                                     1   1%
            - progn                                                 1   1%
             - save-excursion                                       1   1%
              - let                                                 1   1%
               - if                                                 1   1%
                - progn                                             1   1%
                 - setq                                             1   1%
                  - or                                              1   1%
                   - go--in-interface-p                             1   1%
                    - go--in-paren-with-prefix-p                    1   1%
                     - save-excursion                               1   1%
                      - and                                         1   1%
                         string=                                    1   1%
             and                                                    1   1%
        - go--fontify-param-pre                                     1   1%
         - save-excursion                                           1   1%
          - let                                                     1   1%
             if                                                     1   1%
        - go--match-single-func-result                              1   1%
         - let                                                      1   1%
          - while                                                   1   1%
             and                                                    1   1%
        - go--match-type-alias                                      1   1%
         - let                                                      1   1%
          - while                                                   1   1%
           - setq                                                   1   1%
            - or                                                    1   1%
             - go--in-paren-with-prefix-p                           1   1%
              - save-excursion                                      1   1%
               - and                                                1   1%
                - go-goto-opening-parenthesis                       1   1%
                 - let                                              1   1%
                  - nth                                             1   1%
                     syntax-ppss                                    1   1%
- ...                                                              13  20%
 - lsp-organize-imports                                             7  11%
  - condition-case                                                  7  11%
   - lsp-execute-code-action-by-kind                                7  11%
    - let*                                                          7  11%
     - and                                                          7  11%
      - lsp--select-action                                          7  11%
       - -filter                                                    7  11%
        - lsp-get-or-calculate-code-actions                         7  11%
         - lsp-request                                              7  11%
          - let*                                                    7  11%
           - progn                                                  7  11%
            - if                                                    7  11%
             - let*                                                 7  11%
              - unwind-protect                                      7  11%
               - progn                                              7  11%
                - while                                             6   9%
                 - accept-process-output                            1   1%
                  - #<lambda 0xbbef1176e76cf09>                     1   1%
                   - while                                          1   1%
                    - if                                            1   1%
                     - let*                                         1   1%
                      - if                                          1   1%
                       - progn                                      1   1%
                          let                                       1   1%
                - lsp-request-async                                 1   1%
                 - let*                                             1   1%
                  - progn                                           1   1%
                   - lsp--send-request-async                        1   1%
                    - let*                                          1   1%
                     - if                                           1   1%
                      - let*                                        1   1%
                       - seq-do                                     1   1%
                        - mapc                                      1   1%
                         - #<lambda -0x4c538f3a88745f6>                  1   1%
                          - let                                     1   1%
                           - lsp--send-no-wait                      1   1%
                              condition-case                        1   1%
 - #<lambda -0x165fbdd04dda2b6f>                                    5   7%
  - if                                                              5   7%
   - progn                                                          5   7%
    - funcall                                                       5   7%
     - lsp--modeline-update-code-actions                            3   4%
      - setq                                                        3   4%
       - if                                                         3   4%
        - lsp-modeline--build-code-actions-string                   3   4%
         - let*                                                     3   4%
          - lsp-modeline--build-code-actions-segments                  3   4%
           - let*                                                   3   4%
            - mapconcat                                             3   4%
             - #<lambda 0x1f7c00703ae9a93a>                         3   4%
              - cond                                                3   4%
               - lsp-modeline--code-actions-icon                    3   4%
                  if                                                3   4%
     - lsp--apply-text-edit-replace-buffer-contents                  2   3%
      - let*                                                        2   3%
       - let*                                                       2   3%
        - if                                                        1   1%
         - progn                                                    1   1%
          - seq-do                                                  1   1%
           - mapc                                                   1   1%
            - lsp--apply-text-document-edit                         1   1%
             - let*                                                 1   1%
              - cond                                                1   1%
               - let                                                1   1%
                - let*                                              1   1%
                 - if                                               1   1%
                  - save-current-buffer                             1   1%
                   - set-buffer                                     1   1%
                    - find-buffer-visiting                          1   1%
                       file-truename                                1   1%
        - let                                                       1   1%
         - save-current-buffer                                      1   1%
          - unwind-protect                                          1   1%
           - progn                                                  1   1%
            - let                                                   1   1%
             - save-current-buffer                                  1   1%
              - save-excursion                                      1   1%
               - save-restriction                                   1   1%
                - let                                               1   1%
                 - run-hook-with-args                               1   1%
                  - go--reset-dangling-cache-before-change                  1   1%
                     setq                                           1   1%
 - lsp-modeline--check-code-actions                                 1   1%
  - if                                                              1   1%
   - progn                                                          1   1%
    - lsp-request-async                                             1   1%
     - let*                                                         1   1%
      - progn                                                       1   1%
       - lsp--send-request-async                                    1   1%
        - let*                                                      1   1%
         - if                                                       1   1%
          - let*                                                    1   1%
           - seq-do                                                 1   1%
            - mapc                                                  1   1%
             - #<lambda 0x1fa6ae90a678aa0a>                         1   1%
              - let                                                 1   1%
               - lsp--send-no-wait                                  1   1%
                  condition-case                                    1   1%
   Automatic GC                                                     0   0%
```

</details>
<details>
<summary><code>lsp-immediate-organize-imports-first-call-before-pr-2025.txt</code></summary>

Delay did not occurr. Before #2025, first invocation after file save, immediately after modifying buffer content.

```
- command-execute                                                  26  81%
 - byte-code                                                       16  50%
  - read-extended-command                                          16  50%
     completing-read-default                                       12  37%
 - funcall-interactively                                           10  31%
  - execute-extended-command                                        6  18%
     sit-for                                                        4  12%
   - command-execute                                                2   6%
      funcall-interactively                                         2   6%
  - file-notify-handle-event                                        3   9%
   - file-notify--callback-kqueue                                   3   9%
    - file-notify--handle-event                                     3   9%
     - file-notify--call-handler                                    3   9%
      - #<lambda -0x74f39110dd24902>                                3   9%
       - lsp--folder-watch-callback                                 3   9%
        - let                                                       3   9%
         - cond                                                     3   9%
          - let                                                     2   6%
           - let*                                                   2   6%
            - if                                                    2   6%
             - save-current-buffer                                  2   6%
              - let*                                                2   6%
               - if                                                 2   6%
                - lsp--apply-text-edits                             2   6%
                 - if                                               2   6%
                  - let                                             2   6%
                   - unwind-protect                                 2   6%
                    - progn                                         2   6%
                     - prog1                                        2   6%
                      - progn                                       2   6%
                       - let*                                       2   6%
                          lsp--info                                 1   3%
                          unwind-protect                            1   3%
            and                                                     1   3%
    open-line                                                       1   3%
- redisplay_internal (C function)                                   3   9%
 - jit-lock-function                                                2   6%
  - jit-lock-fontify-now                                            2   6%
   - jit-lock--run-functions                                        2   6%
    - #<compiled 0x587239fc7805ef5>                                 2   6%
     - font-lock-fontify-region                                     2   6%
      - font-lock-default-fontify-region                            2   6%
       - font-lock-fontify-syntactically-region                     1   3%
          syntax-ppss                                               1   3%
 - eval                                                             1   3%
  - lsp-mode-line                                                   1   3%
   - let*                                                           1   3%
      if                                                            1   3%
- #<lambda 0x158e992b4c2162d4>                                      2   6%
 - while                                                            2   6%
  - if                                                              2   6%
   - let*                                                           2   6%
    - if                                                            2   6%
     - progn                                                        2   6%
      - if                                                          1   3%
       - let*                                                       1   3%
        - unwind-protect                                            1   3%
         - progn                                                    1   3%
          - while                                                   1   3%
           - accept-process-output                                  1   3%
            - #<lambda 0x158e992b4c2162d4>                          1   3%
             - while                                                1   3%
              - if                                                  1   3%
               - let*                                               1   3%
                - if                                                1   3%
                 - progn                                            1   3%
                  - lsp--parser-on-message                          1   3%
                   - condition-case                                 1   3%
                      let                                           1   3%
      - lsp--parser-on-message                                      1   3%
       - condition-case                                             1   3%
          let                                                       1   3%
- ...                                                               1   3%
 - ->>                                                              1   3%
  - funcall                                                         1   3%
   - ->>                                                            1   3%
    - -map                                                          1   3%
     - -lambda                                                      1   3%
      - #<lambda -0x9a1eba34ab021a4>                                1   3%
       - cond                                                       1   3%
        - let*                                                      1   3%
         - let                                                      1   3%
          - let                                                     1   3%
           - while                                                  1   3%
            - let                                                   1   3%
             - setq                                                 1   3%
              - cons                                                1   3%
               - list                                               1   3%
                  make-symbol                                       1   3%
   Automatic GC                                                     0   0%
```

</details>

**Comment by leungbk** [comment]
Setting `create-lockfiles` to `nil` seems to make the issue go away with the following startup file, which differs by your first `lsp-start-plain` by only that statement:

``` elisp

(require 'package)

(setq debug-on-error t)

(let* ((package-archives '(("melpa" . "https://melpa.org/packages/")
                           ("gnu" . "https://elpa.gnu.org/packages/")))
       (no-byte-compile t)
       (package-user-dir (expand-file-name (make-temp-name "lsp-tmp-elpa")
                                           user-emacs-directory))
       (custom-file (expand-file-name "custom.el" package-user-dir))
       (pkg-list '(lsp-mode company flycheck go-mode)))

  (package-initialize)
  (package-refresh-contents)

  (mapcar (lambda (pkg)
            (unless (package-installed-p pkg)
              (package-install pkg))
            (require pkg))
          pkg-list)

  (setq create-lockfiles nil) ;; <<------
  (add-hook 'prog-mode-hook 'lsp)

  (defun lsp-go-install-save-hooks ()
    (add-hook 'before-save-hook #'lsp-format-buffer t t)
    (add-hook 'before-save-hook #'lsp-organize-imports t t))
  (add-hook 'go-mode-hook #'lsp-go-install-save-hooks)

  (add-hook 'kill-emacs-hook `(lambda ()
                                (delete-directory ,package-user-dir t))))

(provide 'lsp-start-plain)
;;; lsp-start-plain.el ends here
```

The reason it was faster before was that the old glob function we used didn't convert to the correct regexp, preventing a `workspace/didWatchChangedFiles` notification from being sent.

**Comment by leungbk** [comment]
I'm thinking we should simply guard `lsp--file-process-event` with a check to ensure that it always ignores lockfiles.

**Comment by yyoncho** [comment]
@leungbk that is a very good idea. We could skip the backup files as well.
